File: CPlusPlusBowlingGame/tests/BowlingGameTest.cpp
#include <iostream>
#include <cassert>
#include "../src/BowlingGame.hpp"

void testGutterGame() {
    BowlingGame game;
    for (int i = 0; i < 20; ++i) game.roll(0);
    assert(game.score() == 0);
}

void testAllOnes() {
    BowlingGame game;
    for (int i = 0; i < 20; ++i) game.roll(1);
    assert(game.score() == 20);
}

void testOneSpare() {
    BowlingGame game;
    game.roll(5);
    game.roll(5);
    game.roll(3);
    for (int i = 0; i < 17; ++i) game.roll(0);
    assert(game.score() == 16);
}

void testOneStrike() {
    BowlingGame game;
    game.roll(10);
    game.roll(3);
    game.roll(4);
    for (int i = 0; i < 16; ++i) game.roll(0);
    assert(game.score() == 24);
}

void testPerfectGame() {
    BowlingGame game;
    for (int i = 0; i < 12; ++i) game.roll(10);
    assert(game.score() == 300);
}

int main() {
    testGutterGame();
    testAllOnes();
    testOneSpare();
    testOneStrike();
    testPerfectGame();
    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: JavaBowlingGame/src/test/java/BowlingGameTest.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class BowlingGameTest {

    private BowlingGame game;

    @BeforeEach
    public void setUp() {
        game = new BowlingGame();
    }

    private void rollMany(int n, int pins) {
        for (int i = 0; i < n; i++) {
            game.roll(pins);
        }
    }

    private void rollSpare() {
        game.roll(5);
        game.roll(5);
    }

    private void rollStrike() {
        game.roll(10);
    }

    @Test
    public void testGutterGame() {
        rollMany(20, 0);
        assertEquals(0, game.score());
    }

    @Test
    public void testAllOnes() {
        rollMany(20, 1);
        assertEquals(20, game.score());
    }

    @Test
    public void testOneSpare() {
        rollSpare();
        game.roll(3);
        rollMany(17, 0);
        assertEquals(16, game.score());
    }

    @Test
    public void testOneStrike() {
        rollStrike();
        game.roll(3);
        game.roll(4);
        rollMany(16, 0);
        assertEquals(24, game.score());
    }

    @Test
    public void testPerfectGame() {
        rollMany(12, 10);
        assertEquals(300, game.score());
    }
}
